pkg/git/gogit/checkout_test.go

```diff
@@ -359,6 +371,373 @@ func TestCheckoutTagSemVer_Checkout(t *testing.T) {
 	}
 }
 
+// Test_KeyTypes assures support for the different types of keys
+// for SSH Authentication supported by Flux.
+func Test_KeyTypes(t *testing.T) {
+	tests := []struct {
+		keyType ssh.KeyPairType
+	}{
+		{keyType: ssh.RSA_4096},
+		{keyType: ssh.ECDSA_P256},
+		{keyType: ssh.ECDSA_P384},
+		{keyType: ssh.ECDSA_P521},
+		{keyType: ssh.ED25519},
+	}
+
+	serverRootDir := t.TempDir()
+	server := gittestserver.NewGitServer(serverRootDir)
+
+	// Auth needs to be called, for authentication to be enabled.
+	server.Auth("", "")
+
+	var authorizedPublicKey string
+	server.PublicKeyLookupFunc(func(content string) (*gitkit.PublicKey, error) {
+		authedKey := strings.TrimSuffix(string(authorizedPublicKey), "\n")
+		if authedKey == content {
+			return &gitkit.PublicKey{Content: content}, nil
+		}
+		return nil, fmt.Errorf("pubkey provided '%s' does not match %s", content, authedKey)
+	})
+
+	g := NewWithT(t)
+	timeout := 5 * time.Second
+
+	server.KeyDir(filepath.Join(server.Root(), "keys"))
+	g.Expect(server.ListenSSH()).To(Succeed())
+
+	go func() {
+		server.StartSSH()
+	}()
+	defer server.StopSSH()
+
+	repoPath := "test.git"
+	err := server.InitRepo(testRepositoryPath, git.DefaultBranch, repoPath)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	sshURL := server.SSHAddress()
+	repoURL := sshURL + "/" + repoPath
+
+	// Fetch host key.
+	u, err := url.Parse(sshURL)
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(u.Host).ToNot(BeEmpty())
+
+	knownHosts, err := ssh.ScanHostKey(u.Host, timeout, git.HostKeyAlgos)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	for _, tt := range tests {
+		t.Run(string(tt.keyType), func(t *testing.T) {
+			g := NewWithT(t)
+
+			// Generate ssh keys based on key type.
+			kp, err := ssh.GenerateKeyPair(tt.keyType)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			// Update authorized key to ensure only the new key is valid on the server.
+			authorizedPublicKey = string(kp.PublicKey)
+
+			secret := corev1.Secret{
+				Data: map[string][]byte{
+					"identity":    kp.PrivateKey,
+					"known_hosts": knownHosts,
+				},
+			}
+
+			authOpts, err := git.AuthOptionsFromSecret(repoURL, &secret)
+			g.Expect(err).ToNot(HaveOccurred())
+
+			// Prepare for checkout.
+			branchCheckoutStrat := &CheckoutBranch{Branch: git.DefaultBranch}
+			tmpDir := t.TempDir()
+
+			ctx, cancel := context.WithTimeout(context.TODO(), timeout)
+			defer cancel()
+
+			// Checkout the repo.
+			commit, err := branchCheckoutStrat.Checkout(ctx, tmpDir, repoURL, authOpts)
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(commit).ToNot(BeNil())
+
+			// Confirm checkout actually happened.
+			d, err := os.ReadDir(tmpDir)
+			g.Expect(err).ToNot(HaveOccurred())
+			g.Expect(d).To(HaveLen(2)) // .git and foo.txt
+		})
+	}
+
+	// lastly, ensure that if the authentication was not
+	// successful, an error is returned.
```

Comment:
Maybe optional:

Instead of a separate section at the end here, this can be consolidated within the above subtests with a few more fields in the test cases struct, maybe also following the pattern we follow in a lot of our other tests where we run `beforeFunc` just before the function that we are testing and perform any configurations accordingly.

In this case, I think we can update the `tests` struct to something like:

```go
	tests := []struct {
                name string
		keyType ssh.KeyPairType
                beforeFunc func(publicKey []byte)
                wantErr bool
	}
```
`name` can be brought back to describe the case, especially when we have a failure case.

`beforeFunc` can be called before checkout and used to set `authorizedPublicKey`, similar to other usage of `beforeFunc`, for example [this](https://github.com/fluxcd/source-controller/blob/d425923a54efd0f5867aaaeacb46b4ee9a6424ba/controllers/gitrepository_controller_test.go#L489-L491).